pkg/querypb/query_response.proto

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/prompb/types.proto
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: Prometheus Team
+
+syntax = "proto3";
+
+package querypb;
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+message QueryResponse {
+  // These values correspond to the possible status values defined in https://github.com/prometheus/prometheus/blob/main/web/api/v1/api.go.
+  enum Status {
+    option (gogoproto.goproto_enum_prefix) = true;
+    ERROR = 0;
+    SUCCESS = 1;
+  }
+
+  // These values correspond to the possible error type values defined in https://github.com/prometheus/prometheus/blob/main/web/api/v1/api.go.
+  enum ErrorType {
+    option (gogoproto.goproto_enum_prefix) = true;
+    NONE = 0;
+    TIMEOUT = 1;
+    CANCELED = 2;
+    EXECUTION = 3;
+    BAD_DATA = 4;
+    INTERNAL = 5;
+    UNAVAILABLE = 6;
+    NOT_FOUND = 7;
+  }
+
+  Status status = 1;
+  ErrorType error_type = 2;
+  string error = 3;
+
+  oneof data {
+    StringData string = 4;
+    VectorData vector = 5;
+    ScalarData scalar = 6;
+    MatrixData matrix = 7;
+  }
+}
+
+message StringData {
+  string value = 1;
+}
+
+message VectorData {
+  repeated VectorSample samples = 1 [(gogoproto.nullable) = false];
+  repeated VectorHistogram histograms = 2 [(gogoproto.nullable) = false];
+}
+
+message VectorSample {
+  // Why not use a map<...> here? We want to preserve the order of the labels, as labels.Labels expects them to be sorted.
+  repeated string metric = 1;
+
+  double value = 2;
+  int64 timestamp_milliseconds = 3;
+}
+
+message VectorHistogram {
+  // Why not use a map<...> here? We want to preserve the order of the labels, as labels.Labels expects them to be sorted.
+  repeated string metric = 1;
+
+  Histogram histogram = 2 [(gogoproto.nullable) = false];
+}
+
+message ScalarData {
+  double value = 1;
+  int64 timestamp_milliseconds = 2;
+}
+
+message MatrixData {
+  repeated MatrixSeries series = 1 [(gogoproto.nullable) = false];
+}
+
+message MatrixSeries {
+  // Why not use a map<...> here? We want to preserve the order of the labels.
+  repeated string metric = 1;
+
+  repeated MatrixSample samples = 2 [(gogoproto.nullable) = false];
+  repeated Histogram histograms = 3 [(gogoproto.nullable) = false];
+}
+
+message MatrixSample {
+  double value = 1;
+  int64 timestamp_milliseconds = 2;
+}
+
+// This is based on https://github.com/prometheus/prometheus/blob/main/prompb/types.proto
+message Histogram {
```

Comment:
agreed, you should be able to reuse the one in mimir.proto

Reply:
Makes sense, addressed in 8d71462.

Reply:
I've just started looking at implementing conversion from PromQL engine results to this new payload format, and it looks like PromQL will only ever return `histogram.FloatHistogram`s, never `histogram.Histogram`s (as [`promql.Point.H`](https://github.com/prometheus/prometheus/blob/1f0cc09579b7d04bbcb2b9746b7040da55791d48/promql/value.go#L110) is a `FloatHistogram`). Therefore a `QueryResponse` will only ever contain float histograms. 

However, this definition of `Histogram`, which is used for write requests, caters to both float and integer histograms. While we can just ignore all of the integer histogram-specific fields, this could could easily lead to mistakes. (The `oneof` fields are also awkward to work with and wouldn't be needed in a specialised types.) How would you feel about introducing a `FloatHistogram` message that is used by `QueryResponse` @pracucci @krajorama?

Reply:
(I've added a `FloatHistogram` message in adea9e4.)

Reply:
In general I think tests should protect against such mistakes, also the conversion from/to mimirpb.Histogram and (prometheus)histogram.Histogram and histogram.FloatHistogram will be localized into function , e.g. https://github.com/grafana/mimir/blob/3e80f787833ead5ac0a835fc94219ce9f723f0e0/pkg/mimirpb/compat.go#L241  

So I feel it's a bit too redundant and introduces extra code and things to keep in sync. I recommend not doing it.

Reply:
I'll revert it for now, but I'm also keen to benchmark it and see if a specialised type has any performance impact. (This will be for a later PR though.)

Reply:
I've [benchmarked](https://github.com/grafana/mimir/commit/3efffb0b62c0c23d430ae6d8c92e769518209ad9) encoding with the generic `Histogram` type and with a dedicated `FloatHistogram` type, and I think the results justify the extra effort:

```
name                                                                              old time/op    new time/op    delta
ProtobufCodec_Encode/vector_with_float_and_histogram_values-10                      1.03µs ± 0%    0.90µs ± 0%  -12.34%  (p=0.016 n=4+5)
ProtobufCodec_Encode/matrix_with_series_with_histogram_value-10                      535ns ± 0%     482ns ± 0%   -9.84%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_both_float_and_histogram_values-10      862ns ± 0%     757ns ± 0%  -12.11%  (p=0.016 n=4+5)
ProtobufCodec_Encode/vector_with_series_with_histogram_value-10                      493ns ± 0%     427ns ± 0%  -13.40%  (p=0.008 n=5+5)

name                                                                              old alloc/op   new alloc/op   delta
ProtobufCodec_Encode/vector_with_float_and_histogram_values-10                      1.22kB ± 0%    1.06kB ± 0%  -13.07%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_histogram_value-10                       624B ± 0%      544B ± 0%  -12.82%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_both_float_and_histogram_values-10     1.10kB ± 0%    0.94kB ± 0%  -14.49%  (p=0.008 n=5+5)
ProtobufCodec_Encode/vector_with_series_with_histogram_value-10                       536B ± 0%      456B ± 0%  -14.93%  (p=0.008 n=5+5)

name                                                                              old allocs/op  new allocs/op  delta
ProtobufCodec_Encode/vector_with_float_and_histogram_values-10                        17.0 ± 0%      13.0 ± 0%  -23.53%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_histogram_value-10                       10.0 ± 0%       8.0 ± 0%  -20.00%  (p=0.008 n=5+5)
ProtobufCodec_Encode/matrix_with_series_with_both_float_and_histogram_values-10       15.0 ± 0%      11.0 ± 0%  -26.67%  (p=0.008 n=5+5)
ProtobufCodec_Encode/vector_with_series_with_histogram_value-10                       9.00 ± 0%      7.00 ± 0%  -22.22%  (p=0.008 n=5+5)
```

`oneof` requires the use of a pointer value, so with the generic `Histogram` type, we have to perform two allocations per histogram while encoding. With the specialised `FloatHistogram` type, we avoid these allocations and their CPU and heap impact.

Reply:
ack